winml/test/model/model_tests.cpp

```diff
@@ -0,0 +1,166 @@
+      auto actualOutputTensorValue = results.Outputs().Lookup(outputName).as<ITensorNative>();
+      void* actualData;
+      uint32_t actualSizeInBytes;
+      WINML_EXPECT_HRESULT_SUCCEEDED(actualOutputTensorValue->GetBuffer(reinterpret_cast<BYTE**>(&actualData), &actualSizeInBytes));
+
+      // Create a copy of Ort::Value from evaluation output
+      auto expectedShapeAndTensorType = Ort::TensorTypeAndShapeInfo{nullptr};
+      auto memoryInfo = Ort::MemoryInfo{nullptr};
+      WINML_EXPECT_NO_THROW(expectedShapeAndTensorType = it->second.GetTensorTypeAndShapeInfo());
+      WINML_EXPECT_NO_THROW(memoryInfo = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault));
+      Ort::Value actualOutput = Ort::Value{nullptr};
+      WINML_EXPECT_NO_THROW(
+          actualOutput = Ort::Value::CreateTensor(
+              memoryInfo,
+              actualData,
+              actualSizeInBytes,
+              expectedShapeAndTensorType.GetShape().data(),
+              expectedShapeAndTensorType.GetShape().size(),
+              expectedShapeAndTensorType.GetElementType()));
+
+      // Use the expected and actual OrtValues to compare
+      std::pair<COMPARE_RESULT, std::string> ret = CompareOrtValue(*actualOutput, *it->second, m_perSampleTolerance, m_relativePerSampleTolerance, m_postProcessing);
+      WINML_EXPECT_EQUAL(COMPARE_RESULT::SUCCESS, ret.first) << ret.second;
+    }
+  }
+};
+
+TEST_P(ModelTest, Run) {
+  LearningModel model = nullptr;
+  LearningModelDevice device = nullptr;
+  LearningModelSession session = nullptr;
+  LearningModelBinding binding = nullptr;
+  WINML_EXPECT_NO_THROW(model = LearningModel::LoadFromFilePath(m_testCase->GetModelUrl()));
+  WINML_EXPECT_NO_THROW(device = LearningModelDevice(m_deviceKind));
+  WINML_EXPECT_NO_THROW(session = LearningModelSession(model, device));
+  WINML_EXPECT_NO_THROW(binding = LearningModelBinding(session));
+  for (int i = 0; i < m_testCase->GetDataCount(); i++) {
+
+    // Load and bind inputs
+    onnxruntime::test::HeapBuffer inputHolder;
+    std::unordered_map<std::string, Ort::Value> inputFeeds;
+    WINML_EXPECT_NO_THROW(m_testCase->LoadTestData(i, inputHolder, inputFeeds, true));
+    WINML_EXPECT_NO_THROW(BindInputsFromFeed(binding, inputFeeds));
+
+    // evaluate
+    LearningModelEvaluationResult results = nullptr;
+    WINML_EXPECT_NO_THROW(results = session.Evaluate(binding, L"Testing"));
+
+    // Load expected outputs
+    onnxruntime::test::HeapBuffer outputHolder;
+    std::unordered_map<std::string, Ort::Value> outputFeeds;
+    WINML_EXPECT_NO_THROW(m_testCase->LoadTestData(i, outputHolder, outputFeeds, false));
+
+    // compare results
+    CompareEvaluationResults(results, outputFeeds);
+  }
+}
+
+void SetTestDataPath(char* testDataPath) {
+  GetEnvironmentVariableA("WINML_TEST_DATA_PATH", testDataPath, MAX_PATH);
+  if (testDataPath[0] == 0) {
+    throw std::invalid_argument("need to set environment variable WINML_TEST_DATA_PATH");
+  }
+}
```

Comment:
How about `std::string GetTestDataPath()` ?

Reply:
Won't there be a string copy?

Reply:
I want to keep it the same way because this is how we've done it in model tests originally.

Reply:
No, return value optimization applies (it should create the std::string once directly in the output parameter, no copies). The downside is that std::string is allocated in the heap (so there's some kind of `malloc`-like call, which is a bit slower), but shouldn't matter much for test code.

My point was also about the method signature (`void SetTestDataPath(char* testDataPath)` sounds like it's setting the TestDataPath to its argument, something with Get is more intuitive).